### PR TITLE
Move a work item up and down for order of execution (#557)

### PR DESCRIPTION
**Pull request by nimishamukherjee**
This PR will add the UI behaviour only, for moving a work item up or down.
UI Task - #557.

Note: There are two experiences - selecting a work item and checking a work item to move it.
1. Selecting a work item will highlight the work item and open the detailed view.
2. Selecting the checkbox will display the move options menu
3. A work item whose checkbox is selected can be moved up or down
4. If a work item is selected with it's detailed view open, select the checkbox on another work item and move it up or down. The selected work item remains selected and the detailed view remains open.

### CODE REVIEW

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 1]
```diff
@@ -87,6 +87,9 @@
       border-width: 1px 0;
       padding: 1px 10px;
     }
+    .move-dropdown{
+      margin-right: em(1600);
```
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: dropdown is broken. 
screen shot 2017-01-06 at 3 33 52 pm

File: src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.ts [hunk 2]
```diff
@@ -74,6 +76,23 @@ export class WorkItemListEntryComponent implements OnInit {
...
 
+  isChecked(): boolean {
+    return this.chooseWi;
+  }
+
+  check(): void {
+    this.chooseWi = true;
+  }
+
+  decheck(): void {
```
Comment (vikram-raj) [inline src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.ts]: This function name should be unCheck.

File: src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.ts [hunk 3]
```diff
@@ -110,7 +134,7 @@ export class WorkItemListEntryComponent implements OnInit {
       .then(() => {});
   }
 
-  onSelect(event: MouseEvent): void {
+  onSelect(event: MouseEvent): void {
```
Comment (vikram-raj) [inline src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.ts]: Extra space added.

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 4]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
   }
 
   // event handlers
+  onToggle(entryComponent: WorkItemListEntryComponent, event: MouseEvent): void {
+    // This condition is to select a single work item for movement
+    // de-check, de-select, and close detail's panel
```
Comment (vikram-raj) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: de-check should be uncheck. I can't find any word decheck.

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 5]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
   }
 
   // event handlers
+  onToggle(entryComponent: WorkItemListEntryComponent, event: MouseEvent): void {
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: If onToggle accepts two arguments, in **work-item-list.component.html - 95** , why is it being called with one?

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 6]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+    // This condition is to select a single work item for movement
+    // de-check, de-select, and close detail's panel
+    //for prior selected element (if any)
+    if (this.selectedWorkItemEntryComponent && this.selectedWorkItemEntryComponent != entryComponent){
+      this.selectedWorkItemEntryComponent.deselect();
+      this.selectedWorkItemEntryComponent.decheck();
+      this.panelState = 'out';
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: Does changing the panelState to out actually change the URL from details page to list page?

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 7]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+    }
+    // select new component
+    let selected = entryComponent.toggle();
+    let workItem: WorkItem = entryComponent.getWorkItem();
+    if (!selected){
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: Please give space between ) and {

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 8]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+      this.selectedWorkItemEntryComponent.deselect();
+      this.selectedWorkItemEntryComponent = null;
+    }else{
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: Space between } else {

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 9]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+    // This condition is to select a single work item for movement
+    // de-check, de-select, and close detail's panel
+    //for prior selected element (if any)
+    if (this.selectedWorkItemEntryComponent && this.selectedWorkItemEntryComponent != entryComponent){
+      this.selectedWorkItemEntryComponent.deselect();
+      this.selectedWorkItemEntryComponent.decheck();
+      this.panelState = 'out';
+    }
+    // select new component
+    let selected = entryComponent.toggle();
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: This variable is confusing. Is it actually signify selected or checked?

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 10]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+    let workItem: WorkItem = entryComponent.getWorkItem();
+    if (!selected){
+      this.selectedWorkItemEntryComponent.deselect();
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: If I uncheck an item, that not necessarily is to be deselected, it still can have the focus.

File: src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.html [hunk 11]
```diff
@@ -5,6 +5,10 @@
 	></alm-dialog>
 </div>
 <div class="list-group-item" (click)="onSelect($event)" [class.selected]="isSelected()">
+	<!--checkbox to select a WI and move it-->
+	<div class="row-cbh" >
+		<input type="checkbox" [checked]="isChecked()" (click)="toggleEntry($event)" />
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.html]: Why even use this function? I see this function is only used here to get the value of chooseWi, so you can remove this function and use **chooseWi** directly instead.

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 12]
```diff
@@ -175,14 +175,39 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+      this.selectedWorkItemEntryComponent = entryComponent;
+    }
+
+  }
 
   onSelect(entryComponent: WorkItemListEntryComponent): void {
     let workItem: WorkItem = entryComponent.getWorkItem();
     // de-select prior selected element (if any)
-    if (this.selectedWorkItemEntryComponent && this.selectedWorkItemEntryComponent != entryComponent)
+    if (this.selectedWorkItemEntryComponent && this.selectedWorkItemEntryComponent != entryComponent){
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: Need a space between ) and {

File: src/app/work-item/work-item.service.ts [hunk 13]
```diff
@@ -406,16 +407,73 @@ export class WorkItemService {
...
-      this.workItems.splice(this.workItems.length, 0, wi);
+    switch (dir){
+      case 'top':
+        //this.workItems.splice(0, 0, wi);
+        break;
+      case 'bottom':
+        //this.workItems.splice(this.workItems.length, 0, wi);
+        break;
+      case 'up':
+        if (index > 0){ //no moving of element is it is the first element
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: no moving of element **if** it is the first element

Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Need a space between ) and {

File: src/app/work-item/work-item.service.ts [hunk 14]
```diff
@@ -406,16 +407,73 @@ export class WorkItemService {
...
+          //move the work item up by 1. Below statement will create two elements
+          this.workItems.splice( index - 1 , 0, wi);
+          //remove the duplicate element
+          this.workItems.splice( index + 1, 1);
+          //Set the previous and next WI ids
+          if (index -1 === 0){
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: A space between - and 1

Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: You can just write !index - 1

File: src/app/work-item/work-item.service.ts [hunk 15]
```diff
@@ -406,16 +407,73 @@ export class WorkItemService {
...
+            wi.attributes.nextitem = parseInt(this.workItems[index].id);
+          }else{
+            wi.attributes.previousitem = parseInt(this.workItems[index - 2].id);
+            wi.attributes.nextitem = parseInt(this.workItems[index - 1].id);
+          }
+        }
+        break;
+      case 'down':
+        if (this.workItems[ index + 1]) { //no moving of elements if it is the last element
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Check with the length instead. Like, 
**index < (this.workItems.lenght - 1)**
Ask me the reason if you want to know. :)

File: src/app/work-item/work-item.service.ts [hunk 16]
```diff
@@ -406,16 +407,73 @@ export class WorkItemService {
...
+          //move the work item down by 1. Below statement will create two elements
+          this.workItems.splice( index + 2 , 0, wi);
+          //remove the duplicate element
+          this.workItems.splice( index, 1);
+          //Set the previous and next WI ids
+          //Check if the element to move is the second last element or not
+          //and set the previous and next values
+          if ((index + 2) === this.workItems.length) {
+            wi.attributes.previousitem = parseInt(this.workItems[index + 1].id);
+          }else{
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Spaces between } else {

File: src/app/work-item/work-item.service.ts (hunk 14 above)
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Space between ) and {

File: src/app/work-item/work-item.service.ts [hunk 17]
```diff
@@ -406,16 +407,73 @@ export class WorkItemService {
...
+      case 'up':
+        if (index > 0){ //no moving of element is it is the first element
+          //move the work item up by 1. Below statement will create two elements
+          this.workItems.splice( index - 1 , 0, wi);
+          //remove the duplicate element
+          this.workItems.splice( index + 1, 1);
+          //Set the previous and next WI ids
+          if (index -1 === 0){
+            wi.attributes.nextitem = parseInt(this.workItems[index].id);
+          }else{
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Spaces between } else {

File: src/app/work-item/work-item.service.ts [hunk 18]
```diff
@@ -406,16 +407,73 @@ export class WorkItemService {
...
+            wi.attributes.previousitem = parseInt(this.workItems[index - 2].id);
+            wi.attributes.nextitem = parseInt(this.workItems[index - 1].id);
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: This logic is wrong. It'll give you the current work item id not the next one.

File: src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.ts [hunk 19]
```diff
@@ -101,6 +120,11 @@ export class WorkItemListEntryComponent implements OnInit {
     this.selectEvent.emit(this);
   }
 
+  toggleEntry(event: MouseEvent): void{
```
Comment (nimishamukherjee) [inline src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.ts]: No they are not :( no yellow highlights

**Comment by SMahil** [comment]
@nimishamukherjee UI breaks on WI List. Let's have a separate issue for that

**Comment by nimishamukherjee** [comment]
[test]

**Comment by nimishamukherjee** [comment]
[test]

File: src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.scss [hunk 20]
```diff
@@ -5,6 +5,12 @@
   cursor: pointer;
   font-size: rem(12);
   align-items: center;
+  .row-cbh{
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.scss]: A space before {

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 21]
```diff
@@ -175,12 +176,25 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
   }
 
   // event handlers
+  onToggle(entryComponent: WorkItemListEntryComponent): void {
+    // This condition is to select a single work item for movement
+    // deselect the previous checked work item
+    if (this.workItemToMove)
+      this.workItemToMove.uncheck();
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: Can you enclose this line with {}. I understand it's just a single line statement, but enclosing increases readability.

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 22]
```diff
@@ -175,12 +176,25 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+    if (this.workItemToMove == entryComponent){
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: Space between ){

File: src/app/work-item/work-item-list/work-item-list.component.ts [hunk 23]
```diff
@@ -175,12 +176,25 @@ export class WorkItemListComponent implements OnInit, AfterViewInit {
...
+      this.workItemToMove = null;
+    }else{
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.ts]: Spaces between } else and {

File: src/app/work-item/work-item.service.ts [hunk 24]
```diff
@@ -435,16 +436,75 @@ export class WorkItemService {
    * place should be implemented
    */
   moveItem(wi: WorkItem, dir: String) {
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Add the return type as **void** or **Promise<void>** for future.

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 25]
```diff
@@ -87,6 +87,10 @@
       border-width: 1px 0;
       padding: 1px 10px;
     }
+    .move-dropdown{
```
Comment (vikram-raj) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: space between .move-dropdown and {

**Comment by vikram-raj** [comment]
@nimishamukherjee Add cursor pointer property to move dropdown menu.

File: src/app/work-item/work-item.service.ts [hunk 26]
```diff
@@ -435,16 +436,75 @@ export class WorkItemService {
...
+        break;
+      case 'up':
+        if (index > 0) { //no moving of element if it is the first element
+          //move the work item up by 1. Below statement will create two elements
+          this.workItems.splice( index - 1 , 0, wi);
+          //remove the duplicate element
+          this.workItems.splice( index + 1, 1);
+          //Set the previous and next WI ids
+          if (index === 1) {
+            wi.attributes.nextitem = parseInt(this.workItems[index].id);
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: This line and 460 are same and looks repetation. So, you can replace following code with 456 to 461.
```
wi.attributes.nextitem = parseInt(this.workItems[index].id);
if (index !== 1) {
    wi.attributes.previousitem = parseInt(this.workItems[index - 2].id);
}
```

File: src/app/work-item/work-item.service.ts [hunk 27]
```diff
@@ -435,16 +436,75 @@ export class WorkItemService {
...
+          this.workItems.splice( index + 2 , 0, wi);
+          //remove the duplicate element
+          this.workItems.splice( index, 1);
+          //Set the previous and next WI ids
+          //Check if the element to move is the second last element or not
+          //and set the previous and next values
+          if ((index + 2) === this.workItems.length) {
+            wi.attributes.previousitem = parseInt(this.workItems[index + 1].id);
+          } else {
+            wi.attributes.previousitem = parseInt(this.workItems[index + 2].id);
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: This should be the nextitem and the other one should be previousitem. Also you can write the entire logic as the one I mentioned above.

File: src/app/work-item/work-item.service.ts [hunk 28]
```diff
@@ -435,16 +436,75 @@ export class WorkItemService {
...
+        if ( index < (this.workItems.length - 1) ) { //no moving of elements if it is the last element
+          //move the work item down by 1. Below statement will create two elements
+          this.workItems.splice( index + 2 , 0, wi);
+          //remove the duplicate element
+          this.workItems.splice( index, 1);
+          //Set the previous and next WI ids
+          //Check if the element to move is the second last element or not
+          //and set the previous and next values
+          if ((index + 2) === this.workItems.length) {
+            wi.attributes.previousitem = parseInt(this.workItems[index + 1].id);
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: This should be **index** instead of **index + 1**

**Comment by nimishamukherjee** [comment]
[test]

**Comment by nimishamukherjee** [comment]
@vikram-raj added cursor for the drop down menu items

**Pull request merged**